New Vehicle Lights Feature - RE: #1820

jtsage opened the pull request:
# New Settings Pages

![AdSettingsPage](https://user-images.githubusercontent.com/605986/133291331-116c9cb6-1f51-46e7-b52d-e866298dfc34.png)
![AdVehicleSettingsPage](https://user-images.githubusercontent.com/605986/133291334-c3f46559-9d02-4c0e-8582-6a1449b85d42.png)

# Global Options

## Inclement Weather _(A)_

This options turns on the usual "night" lighting lights during the day when the weather is inclement (rain or snow).  It respects the light state options from __E__ & __F__

## Time to Turn on and off _(B & C)_

Setting these options to a value other than "--" _(Note: I couldn't figure how to translate only the first entry of the settings list, and there are far to many to want to provide l10n for all of them, so "--" instead of something like "Do not override")_

The time to turn on is a populated list that runs from 16:00 (4p) to 23:30 (11:30p), in half hour increments

The time to turn off is a populated list that runs from 00:30 (12:30a) to 10:00 (10a), in half hour increments

# Per-Vehicle Options

## Daytime Running Lights _(D)_

Toggle this setting to use the first light state on during the daytime. For vanilla vehicles, this is usually the standard headlights, for mod vehicles, this is often the fancy running lights.

## Off Field State _(E)_

This sets the "state" for the lights, at night, when not on the field.  Please see below for an explanation of light states

## Load / Unload State _(F)_

This sets the "state" for the lights, at night, when approaching a (in use) load or unload trigger.  My brief testing supports that it works on pickup / deliver, deliver only, and load type trips.

## Hazards Off Field _(G)_

This turns the hazard lights (both turn signals) on when not on the field (all times of the day).  It disables "Turnlight Automatic" for the vehicle as well.

# What is a light "state"

Light states are available from the vehicle XML in the form of:

```xml
<lights>
     <!- ... ->
    <states>
        <state lightTypes="0" />
        <state lightTypes="0 1" />
        <state lightTypes="0 1 2" />
    </states>
    <!- ... ->
</lights>
```

This allows us to use the light sets / light types that the vehicle designer intended, it also keeps us from assuming that the vehicle uses a ```lightType``` in the same way as the vanilla vehicles and the documentation suggest _(Most of the truck mods I have do not, they alter the spec to provide things like pretty running lights)_

All uses of light states do have sanity checks around them so that if the requested state does not exist, a sane default will be used instead (most often, ```lightsTypeMask = 1```, which turns on ```lightType = 0```

# Provided without an associated option

Backup warning lights should turn on _(if available on the vehicle, in my testing only the vanilla trucks and cars seem to have them, most tractors do not)_ anytime AD is driving in reverse.

# Still missing from #1820 

The only bit that I did not really address is the BGA trigger lights, as I have never used AD for BGA operation, and I don't really know how to set up a nice test case.  I would note however that if you are loading a bunker using a deliver route, the setting above for load/unload should function _(F)_.  The same would be true when any unloader approaches the BGA trigger, assuming that is handled the same way that load/unload usually is.  The only missing bit would be when scooping silage out of the bunker. _(Maybe, again, I've never used AD like this, not sure how to test)_

I've made no changes to how beacon lights are handled, I wasn't sure what, if any, additional options are needed / wanted for that. Right now, beacon lights can be in use when not on the field.

I made no changes to how on the field lights are handled, which is set to use ```aiLightsTypesMask```, which is computed by the vehicle loader, and translates to ```lightTypes``` 0, 1, & 2, if they exist, unless it has been overridden by the vehicle.

# Testing

I've done a fair bit of testing with these changes in, so far I've noticed no oddities, performance issues, and the game log is still clean.  I will continue to play with these changes keeping an eye out for any problems. _(Note: I have never been a multi-player person, so I cannot speak to any problems there)_

# Subset of options

I did try to keep one option per commit in this pull request, so if there is a feature that is not suitable, it should be a simple matter to back out those changes.

jtsage committed:
Feature Add: "lightsOnRain"

Global toggle to use lights when it is raining.

jtsage committed:
Feature Add: "lightsDaytimeRunning"

Per vehicle setting to use the first light style as daytime running lights (equivalent to hitting the lights key once).  Sanity trap: if the styles are somehow not defined, use lightsMask(1) (lightType = 0)

jtsage committed:
Feature Add: "lightsOffFieldState"

Configue the light state to use off-field on a per-vehicle basis.  Sanity check: if the state does not exist for the vehicle, default to a lightMask of "1" (lighType = 0)

jtsage committed:
Picky terminolgy, icon change

This commit should be merged back.

jtsage committed:
Move "lightsOnRain" to where it really should be.

jtsage committed:
Add feature: "reverse driving warning lights"

This one is more complex really, as it appears *most* of the stock tractors don't have reverse lights.  So, should we be using rear worklights, at least at night?  To be discussed.

jtsage committed:
Add Feature "lightsHazardsOffField"

When not on the field, allow per-vehicle setting to run with hazard lights on (both turn signals) - Pretty common to see large equipment run down streets like this, at least in the US.  This should override blinkangle correctly off-field, and allow it to still work on-field.  Testing of this to follow.

jtsage committed:
Add Features "lightsTurnOfTime" and "lightsTurnOffTime"

Allow the user to override the map/seasons/giants default of when headlights should turn on and off.  Pre-poulated off time to be between 00:30 (12:30a) and 10:00 (10a), and on time between 16:00 (4p) and 23:30 (11:30p).  Avoided 00:00 and 24:00 because I have no idea how rounding would affect things, but if wanted, we could internally use 00:01 (1) and 23:59 (1439) and avoid any posible issue.

jtsage committed:
Didn't like title for weather setting, fix off field light state == 0 (no lights)

As to the second, don't know why anybody would want it, but the option was already there from earlier, it just didn't work.

jtsage committed:
Add Feature: "lightsLoadUnloadState"

This lets the user choose a per-vehicle state of the lights when approaching a load or unload trigger (that will be used)

jtsage committed:
"lightsLoadUnloadState" - poor default for vanilla.

A Fair few of the vanilla tractors seem to have only 2 light states - front, front & back.  Assume lowest common denominator, as it is per-vehicle anyway

Axel32019 commented:
Hallo @jtsage, Thanks for your contribution here!
Much text to read, may be I miss something - Sorry to much to do...
In general you should always focus on ingame vehicles when working on a mod.
There are too many strange mods outside to get all of them working!

Here are my findings / hints / ideas...

1. Settings.lua
If you use a text and/or tooltip with language content, it should include translate = true
If a setting is vehicle specific or global it is always good to define isVehicleSpecific = true/false
default should always be the same as current => looks OK!
Have you selected default values which represent an unchanged behaviour compared to the base game?
Otherwise users might be confused about what happen now (i.e. lightsOffFieldState, lightsLoadUnloadState...).

lightsTurnOnTime / lightsTurnOffTime
From my personal impression: 
A resolution of 0:30 might be to much
1:00 with extended range from 00:00 to 12:00 and 12:00 to 23:59 would cover the hole day
What do you think about this?

lightsDaytimeRunning
Would it be consequent to use the same setting values here as well, "lights key", ?

lightsHazardsOffField
Is it in practical from your experience the case that hazards are turned on while reverse driving?

For the title of the buttons we might need find a shorter text to be completely readable.


2. Specialization.lua - AutoDrive:updateAILights
Be aware of the conditions:
AutoDrive:updateAILights is called by the engine cyclic. 
if self.ad ~= nil and self.ad.stateModule:isActive() then
Is only true during AD is active - see also below to switch off...

(AutoDrive.getSettingState("lightsTurnOffTime") > 1) and AutoDrive.getSetting("lightsTurnOffTime")
Is this OK, or swapped?

3. DrivePathModule.lua - ADDrivePathModule:checkActiveAttributesSet
This function is not called all the time if reverse driving, so this will not completely match -> TODO
See ADDrivePathModule:update

I figured out that enabled light, i.e. drive light, stay ON after AD switched OFF -> TODO
The following is one of many performance improvements:
if ((g_updateLoopIndex + self.vehicle.id) % AutoDrive.PERF_FRAMES == 0) then
This conditions is only true during 1 frame out of N. Means AD could be deactivated without hitting this condition!
Every enabled light should be switched of / set to default if AD is deactivated -> TODO
You may have a look at if setBeaconLightsVisibility how to do...

Your open questions:
Lights on before trigger - is there a section / place in the code where I can look at how AD knows it's approaching a trigger?
A utility function maybe? And what criteria for this - lights on when approaching a fill trigger only when the vehicle is going to use it?
I think you already found isInRangeToLoadUnloadTarget, right?

Lights for the BGA - i found AutoDrive.isVehicleInBunkerSiloArea(vehicle) but it looks like that will only return useful (to me) results if the vehicle is actively filling the bunker. 
It wouldn't help for unloading if I am reading this correctly.
It might be neccessary to add this in ADDrivePathModule:followWaypoints, see "vehicle enters drive through bunker silo", but here only for drive-through silo
The reverse silo could be found here: ADSpecialDrivingModule:handleReverseDriving

My idea is to collect all these non-functional settings on a separate page, the already available getting filled more and more, which confuse users with each new setting.
I will discuss this in the team and we will see where to add the new settings.

jtsage commented:
First off, thank you for your feedback, this is excellent - I'll attempt to reply inline, because as you mentioned, this is a lot of stuff. :)


> Hallo @jtsage, Thanks for your contribution here!
> Much text to read, may be I miss something - Sorry to much to do...
> In general you should always focus on ingame vehicles when working on a mod.
> There are too many strange mods outside to get all of them working!
Absolutely.  For reference, my testing has been picking a random tractor from the base game, and also a random choice of truck cabs from the base game (the only vehicles besides the stuff in "cars" that seem to have reverse warning lights defined).
> 
> Here are my findings / hints / ideas...


> 1. Settings.lua
>    If you use a text and/or tooltip with language content, it should include translate = true

translate = false is currently not turned on for the light state options (numeric options only), and the timeOn / timeOff settings (time options only, and a "--" to user the map / seasons / game default).  I didn't see an option to "mix in" translated and not translated text - If I translate timeOn/timeOff, it would be nice for that first option to read "Default", but that would (I think) make translating 01:00, 02:00, etc necessary, which is just extra lines in the l10n files that won't ever change.

>    If a setting is vehicle specific or global it is always good to define isVehicleSpecific = true/false

Got it, I'll tag the few global I missed as isVehicleSpecific = false

>    default should always be the same as current => looks OK!
>    Have you selected default values which represent an unchanged behaviour compared to the base game?
>    Otherwise users might be confused about what happen now (i.e. lightsOffFieldState, lightsLoadUnloadState...).

Where possible, yes.
 * Inclement weather (rain/snow) - not sure if the base game AI turns on lights for rain.  I suspect it doesn't, I will test, but I did default this to ON (following CP's lead)
 * Backup lights - I'm not sure if the base AI uses these or not, so few of the vanilla vehicles actually have backup warning lights.
 * Off Field lights - not applicable, the base AI doesn't have an equivalent feature
 * Lights load/unload - again, not applicable, base AI doesn't have an equivalent - the default I have in there turns on light state "2", which is a mixed bag for the vanilla vehicles - for the smaller vanilla vehicles, this is all the lights, larger ones it's headlights and back works. It's harder to pick a default here, as it won't be the logical choice for all vehicles.  Maybe I should set this back to light state "1", which would be the equivalent of not really changing anything?


> lightsTurnOnTime / lightsTurnOffTime
> From my personal impression:
> A resolution of 0:30 might be to much
> 1:00 with extended range from 00:00 to 12:00 and 12:00 to 23:59 would cover the hole day
> What do you think about this?

I'm not opposed at all. For the upper and lower bounds, I'll display 00:00 and 24:00, in reality, it will be 00:01 and 23:59 - I don't expect that the engine would have an issue with the real values, but from my playing with dates/times in many projects, I have learned to stay away from day boundaries if at all possible.

To take the next step, if you set offTime to 12:00 and onTime to 12:00, the lights would _always_ be on - does an option to explicitly do this make sense? These times are not vehicle specific, because I believe that would make changing them an absolute nightmare of a chore, but maybe an "always on" option per vehicle does make some sense? (I'm trying to think of a scenario - the most obvious to me would be flicking lights on in a forktruck to let anyone else in the yard know it's moving - not really an AD task, but maybe there is a good use case here?)

> lightsDaytimeRunning
> Would it be consequent to use the same setting values here as well, "lights key", ?

Yeah, that's not a bad idea at all.  I've yet to use a vehicle where it would make visual sense to use anything other than state "1", but everyone's play style is different.


> lightsHazardsOffField
> Is it in practical from your experience the case that hazards are turned on while reverse driving?

This is probably a very good idea - it would likely make the most sense to create a per-vehicle setting for reverse driving lights, and allow a few different options - this would also address the issue that most of the vanilla vehicles do not actually have reverse warning lights.

* __reverseDrivingLights__
   * None 
   * Warning Lights
   * Hazard Lights
   * Warning and Hazard Lights

It would be possible to add the option of using rear work lights, but we run into the problem of not knowing which lightType actually *is* the rear lights.  Giants spec say one thing, and most of the vanilla vehicles adhear to this, but mod authors do their own thing.  If I just list the lightTypes, that assumes the user has a certain level of knowledge of how the mod is set up, which doesn't seem fair either.  Using light states like I have elsewhere doesn't really solve it either, as I'd want *just* the rear work lights for backup, not the headlights and rear works (during the day anyways)
   

> For the title of the buttons we might need find a shorter text to be completely readable.

Yes.  My input on this a bit further down.

> 
> 1. Specialization.lua - AutoDrive:updateAILights
>    Be aware of the conditions:
>    AutoDrive:updateAILights is called by the engine cyclic.
>    if self.ad ~= nil and self.ad.stateModule:isActive() then
>    Is only true during AD is active - see also below to switch off...
> 
> (AutoDrive.getSettingState("lightsTurnOffTime") > 1) and AutoDrive.getSetting("lightsTurnOffTime")
> Is this OK, or swapped?

```lua
local nightEnd = (AutoDrive.getSettingState("lightsTurnOffTime") > 1) and AutoDrive.getSetting("lightsTurnOffTime") or g_currentMission.environment.nightEndMinutes
```

This is OK, this is as close as LUA gets to a real ternary operator - if lightsTurnOffTime (OnTime too) is the first setting state, it's set to "use Map / Seasons / Game defaults [however the engine chooses precedence]", otherwise use the AD user configured start/end time. Functionally equivalent to:
```lua
local nightEnd = g_currentMission.environment.nightEndMinutes

if AutoDrive.getSettingState("lightsTurnOffTime") > 1 then
    nightEnd = AutoDrive.getSetting("lightsTurnOffTime")
end
```

> 
> 1. DrivePathModule.lua - ADDrivePathModule:checkActiveAttributesSet
>    This function is not called all the time if reverse driving, so this will not completely match -> TODO
>    See ADDrivePathModule:update

I see what you mean, yeah, let me look at this a bit more

> I figured out that enabled light, i.e. drive light, stay ON after AD switched OFF -> TODO
> The following is one of many performance improvements:
> if ((g_updateLoopIndex + self.vehicle.id) % AutoDrive.PERF_FRAMES == 0) then
> This conditions is only true during 1 frame out of N. Means AD could be deactivated without hitting this condition!
> Every enabled light should be switched of / set to default if AD is deactivated -> TODO
> You may have a look at if setBeaconLightsVisibility how to do...

Do you mean in ```AutoDrive:stopAutoDrive()```?  I hadn't added anything in there to shut down any other lights, as I wasn't sure what the intended behavior was - right now, from my observations, if AutoDrive finishes, and you are not in the vehicle, the lights and motor turn off. If you are in the vehicle, lights and motor stays on.  In both cases, assuming the beacon lights setting is on, they turn off.  The same is true of the "hazardsOffField" setting, as there is a bit in there that turns off any active turn signals.

> 
> Your open questions:
> Lights on before trigger - is there a section / place in the code where I can look at how AD knows it's approaching a trigger?
> A utility function maybe? And what criteria for this - lights on when approaching a fill trigger only when the vehicle is going to use it?
> I think you already found isInRangeToLoadUnloadTarget, right?

Yes, correct.  This covers BGA loading too, assuming that you are using a deliver or load & deliver route.  I don't have a bunker on my test map, so I will double check, but with how those route work with the BGA, the load/unload light state should come on right as unloading the trailer starts.

> 
> Lights for the BGA - i found AutoDrive.isVehicleInBunkerSiloArea(vehicle) but it looks like that will only return useful (to me) results if the vehicle is actively filling the bunker.
> It wouldn't help for unloading if I am reading this correctly.
> It might be neccessary to add this in ADDrivePathModule:followWaypoints, see "vehicle enters drive through bunker silo", but here only for drive-through silo
> The reverse silo could be found here: ADSpecialDrivingModule:handleReverseDriving

Got it, I'll look closer at this. I think maybe some discussion should be had about what is the most appropriate use of lights for BGA operation - does it make sense to switch between forward and back work lights depending on direction, or is it more accurate that the farmer would turn on all work lights so they could see what they are doing?  I honestly have neither any idea, nor opinion, as I am very, very far from being a real farmer. 


> My idea is to collect all these non-functional settings on a separate page, the already available getting filled more and more, which confuse users with each new setting.
> I will discuss this in the team and we will see where to add the new settings.

I really like this idea - A little tab at the top with a light bulb would be nice.  It would also fix the problem you mentioned above that the settings title are way to long.  If I can add my $0.02, I would think that beacon lights should be in this page as well.

My largest hesitation of adding an onFieldLightState (to override aiLightsTypesMask if wanted) was that the settings page was getting mighty full already. In my mind, this would have a setting of ["Default (aiLightsTypesMask)", 0,1,2,3,4,5]



Thanks for the input - if I missed anything on my run through this, please let me know.  For what it's worth, I'm also in the official discord, same user name if for some reason that is easier.

jtsage committed:
Missed "isVehicleSpecific" option.

jtsage committed:
lightsTurnOnTime, lightsTurnOffTime granularity change

Change to 1hr granularity, run turnOffTime from 00:00 to 12:00, and turnOnTime 12:00 to 24:00

Axel32019 commented:
Hallo @jtsage,
Thank you for the replies! I know you are in discord as well, appreciate your support there.
Thinking about the topic, I am more and more afraid about the spreading of all the lights switching in the AD-code, so it might be an idea to summarize all in one single function.
This will give 1 point the implement it all and not forget something to switch On, Off, Default etc.
